### PR TITLE
Offers: Part 3.  Offers and the plugins who love them!

### PR DESCRIPTION
**Pull request by rustyrussell**
This contains plugins both for creating offers and servicing invoice requests / invoices associated with them, and sending requests based on others' offers.

Changelog-None

### CODE REVIEW

**Comment by PapauloGamerOfc** [comment]
The Travis CI bot is falining on every PR!

File: hsmd/hsmd.c [hunk 1]
```diff
@@ -490,6 +491,38 @@ static void populate_secretstuff(void)
...
+	 * We encourage different schemes to apply for assigning a separate
+	 * BIP number and use the same number for purpose field, so addresses
+	 * won't be generated from overlapping BIP32 spaces.
+	 *
+	 * Example: Scheme described in BIP44 should use 44' (or 0x8000002C)
+	 * as purpose.
+	 */
+	/* Clearly, we should use 9735, the unicode point for lightning! */
+	if (bip32_key_from_parent(&master_extkey,
+				  BIP32_INITIAL_HARDENED_CHILD|9735,
```
Comment (cdecker) [inline hsmd/hsmd.c:514]: Pull the constant into a `#define`

File: lightningd/hsm_control.c [hunk 2]
```diff
@@ -125,7 +125,8 @@ struct ext_key *hsm_init(struct lightningd *ld)
 	bip32_base = tal(ld, struct ext_key);
 	msg = wire_sync_read(tmpctx, ld->hsm_fd);
 	if (!fromwire_hsmd_init_reply(msg,
-				     &ld->id, bip32_base)) {
+				      &ld->id, bip32_base,
+				      &ld->bolt12_base)) {
```
Comment (cdecker) [inline lightningd/hsm_control.c:129]: Seems not to be initialized anywhere.

File: hsmd/hsmd.c (hunk 1 above)
Comment (rustyrussell) [reply hsmd/hsmd.c:514]: It's literally used in a single place though; moving the definition makes it harder to understand, not easier.

File: lightningd/hsm_control.c (hunk 2 above)
Comment (rustyrussell) [reply lightningd/hsm_control.c:129]: ?  This is fromwire?

Comment (cdecker) [reply lightningd/hsm_control.c:129]: Right, getting cross-eyed :-)

**Comment by rustyrussell** [comment]
Rebased on master since part 2 committed, no other changes.

File: tests/test_pay.py [hunk 3]
```diff
@@ -3842,3 +3842,27 @@ def test_offer(node_factory, bitcoind):
...
+
+    # Single-use invoice can be fetched multiple times, only paid once.
+    offer = l3.rpc.call('offer', {'amount': '1msat',
+                                  'description': 'single-use test',
+                                  'single_use': True})['bolt12']
+    print(offer)
+
+    inv1 = l1.rpc.call('fetchinvoice', {'offer': offer})
+    inv2 = l1.rpc.call('fetchinvoice', {'offer': offer})
+    assert inv1 != inv2
```
Comment (cdecker) [inline tests/test_pay.py:3872]: Paying and then fetching should fail.

**Review by cdecker — APPROVED** [review]
ACK https://github.com/ElementsProject/lightning/pull/4257/commits/f553635346721632f511cc4f2841de643c4389cc

**Comment by cdecker** [comment]
Fixed the two typos and rebased on `master` ([range-diff](http://46.101.246.115:8888/ElementsProject/lightning/pull/4257/range_diff/f553635346721632f511cc4f2841de643c4389cc..d2e4c5cf893d398fc3c167a1264fbb4885dc81f7).

ACK https://github.com/ElementsProject/lightning/pull/4257/commits/d2e4c5cf893d398fc3c167a1264fbb4885dc81f7

**Pull request merged**